examples/java8/src/main/java/org/apache/beam/examples/advanced/subprocess/utils/CallingSubProcessUtils.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.examples.advanced.subprocess.utils;
+
+import com.google.common.collect.Sets;
+
+import java.util.Set;
+import java.util.concurrent.Semaphore;
+
+import org.apache.beam.examples.advanced.subprocess.configuration.SubProcessConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class for dealing with concurrency and binary file copies to the worker.
+ */
+public class CallingSubProcessUtils {
+
+  // Prevent Instantiation
+  private CallingSubProcessUtils() {};
+
+  static final Logger LOG = LoggerFactory.getLogger(CallingSubProcessUtils.class);
+
+  static boolean initCompleted = false;
+
+  // Allow multiple subclasses to create files, but only one thread per subclass can add the file to
+  // the worker
+  private static final Set<String> downloadedFiles = Sets.<String>newConcurrentHashSet();
+
+  // Limit the number of threads able to do work
+  private static Semaphore semaphore;
```

Comment:
This will end up being a global limit. Say you have different transforms executing completely different subprocesses they will create different instances of these class. They will probably think they are independent, but they won't be - the concurrency limit will apply to the sum of both of them.

Reply:
That is actually desirable as any DoFn that makes use of the semaphores is by definition a 'heavy' process that the user would want to limit concurrency on a worker for.